reconcile/kafka_clusters.py

```diff
@@ -0,0 +1,116 @@
+import sys
+import logging
+import semver
+import base64
+
+import reconcile.queries as queries
+import reconcile.openshift_base as ob
+
+from utils.openshift_resource import OpenshiftResource as OR
+from utils.ocm import OCMMap
+from utils.defer import defer
+
+QONTRACT_INTEGRATION = 'kafka-clusters'
+QONTRACT_INTEGRATION_VERSION = semver.format_version(0, 1, 0)
+
+
+def construct_oc_resource(data):
+    body = {
+        "apiVersion": "v1",
+        "kind": "Secret",
+        "type": "Opaque",
+        "metadata": {
+            "name": "kafka",
+        },
+        "data": {
+            k: base64.b64encode(v.encode()).decode('utf-8')
+            for k, v in data.items()
+        }
+    }
+    return OR(body, QONTRACT_INTEGRATION, QONTRACT_INTEGRATION_VERSION)
+
+
+def fetch_desired_state(clusters):
+    desired_state = []
+    for cluster_info in clusters:
+        item = {
+            'name': cluster_info['name'],
+            'cloud_provider': cluster_info['spec']['provider'],
+            'region': cluster_info['spec']['region']
+        }
+        desired_state.append(item)
+    return desired_state
+
+
+@defer
+def run(dry_run, thread_pool_size=10,
+        internal=None, use_jump_host=True, defer=None):
+    settings = queries.get_app_interface_settings()
+    kafka_clusters = queries.get_kafka_clusters()
+    ocm_map = OCMMap(clusters=kafka_clusters,
+                     integration=QONTRACT_INTEGRATION,
+                     settings=settings)
+    namespaces = []
+    for kafka_cluster in kafka_clusters:
+        namespaces.extend(kafka_cluster['namespaces'])
+    ri, oc_map = ob.fetch_current_state(
+        namespaces=namespaces,
+        thread_pool_size=thread_pool_size,
+        integration=QONTRACT_INTEGRATION,
+        integration_version=QONTRACT_INTEGRATION_VERSION,
+        override_managed_types=['Secret'],
+        internal=internal,
+        use_jump_host=use_jump_host)
+    defer(lambda: oc_map.cleanup())
+
+    current_state = ocm_map.kafka_cluster_specs()
+    desired_state = fetch_desired_state(kafka_clusters)
+
+    error = False
+    for kafka_cluster in kafka_clusters:
+        kafka_cluster_name = kafka_cluster['name']
+        desired_cluster = [c for c in desired_state
+                           if kafka_cluster_name == c['name']][0]
+        current_cluster = [c for c in current_state
+                           if kafka_cluster_name == c['name']]
+        # check if cluster exists. if not - create it
+        if not current_cluster:
+            logging.info(['create_cluster', kafka_cluster_name])
+            if not dry_run:
+                ocm = ocm_map.get(kafka_cluster_name)
+                ocm.create_kafka_cluster(desired_cluster)
```

Comment:
We probably have to do error handling here...
```
[2020-11-11 16:54:32] [DEBUG] [connectionpool.py:_make_request:391] - https://api.stage.openshift.com:443 "POST /api/managed-se
rvices-api/v1/kafkas?async=true HTTP/1.1" 400 203
[2020-11-11 16:54:32] [ERROR] [ocm.py:_post:505] - {"id":"8","kind":"Error","href":"/api/managed-services-api/v1/errors/8","cod
e":"MGD-SERV-API-8","reason":"region us-west-1 is not supported for aws, supported regions are: [us-east-1]","operation_id":"1k
9XEtrscyNpdPXjMesFYc1dRRN"}
Traceback (most recent call last):
  File "/home/apahim/git/qontract-reconcile/venv/bin/qontract-reconcile", line 33, in <module>
    sys.exit(load_entry_point('reconcile', 'console_scripts', 'qontract-reconcile')())
  File "/home/apahim/git/qontract-reconcile/venv/lib64/python3.9/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/home/apahim/git/qontract-reconcile/venv/lib64/python3.9/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/home/apahim/git/qontract-reconcile/venv/lib64/python3.9/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/apahim/git/qontract-reconcile/venv/lib64/python3.9/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/apahim/git/qontract-reconcile/venv/lib64/python3.9/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/home/apahim/git/qontract-reconcile/utils/binary.py", line 14, in f_binary
    f(*args, **kwargs)
  File "/home/apahim/git/qontract-reconcile/venv/lib64/python3.9/site-packages/click/decorators.py", line 21, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/home/apahim/git/qontract-reconcile/reconcile/cli.py", line 1100, in kafka_clusters
    run_integration(reconcile.kafka_clusters, ctx.obj, thread_pool_size,
  File "/home/apahim/git/qontract-reconcile/reconcile/cli.py", line 335, in run_integration
    func_container.run(dry_run, *args, **kwargs)
  File "/home/apahim/git/qontract-reconcile/utils/defer.py", line 15, in func_wrapper
    return func(*args, defer=defer, **kwargs)
  File "/home/apahim/git/qontract-reconcile/reconcile/kafka_clusters.py", line 81, in run
    ocm.create_kafka_cluster(desired_cluster)
  File "/home/apahim/git/qontract-reconcile/utils/ocm.py", line 492, in create_kafka_cluster
    self._post(api, data)
  File "/home/apahim/git/qontract-reconcile/utils/ocm.py", line 506, in _post
    raise e
  File "/home/apahim/git/qontract-reconcile/utils/ocm.py", line 503, in _post
    r.raise_for_status()
  File "/home/apahim/git/qontract-reconcile/venv/lib64/python3.9/site-packages/requests/models.py", line 940, in raise_for_stat
us
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 400 Client Error: Bad Request for url: https://api.stage.openshift.com/api/managed-services-api/
v1/kafkas?async=true
```

Reply:
i've added a region validation in https://gitlab.cee.redhat.com/service/app-interface/-/merge_requests/11847.
i don't want to handle errors yet, i want us to see all the errors for this new integration (and service). in the future, i believe they will add preflight checks that we will be able to use as a preventive measure. wdyt?

Reply:
soungs like a plan :)